### ISSUE TITLE
AesExtraDataRecord not present in local header for aes encrypted data

### ISSUE THREAD

**Issue by jrbocarsly** [srikanth-lingala/zip4j#248]
This occurs in (apparently large) zip files created with zip4j (v2.2.4) on attempting to open the zip. Size is ~1.6 GB. 
We tried to extract files from the zip with 7zip; some files could be extracted and some could not (7zip gives a 'bad password' error when the file cannot be extracted). We also tried with zip4j 2.6.3 and a similar error occurred. We're happy to share the sample zip file once suitable means can be arranged.

Stacktrace:
java.lang.RuntimeException: AesExtraDataRecord not present in local header for aes encrypted data
        at net.lingala.zip4j.util.Zip4jUtil.getCompressionMethod(Zip4jUtil.java:96) ~[zip4j-2.2.4.jar:na]
        at net.lingala.zip4j.io.inputstream.ZipInputStream.getCompressedSize(ZipInputStream.java:290) ~[zip4j-2.2.4.jar:na]
        at net.lingala.zip4j.io.inputstream.ZipInputStream.initializeEntryInputStream(ZipInputStream.java:199) ~[zip4j-2.2.4.jar:na]
        at net.lingala.zip4j.io.inputstream.ZipInputStream.getNextEntry(ZipInputStream.java:106) ~[zip4j-2.2.4.jar:na]
        at net.lingala.zip4j.io.inputstream.ZipInputStream.getNextEntry(ZipInputStream.java:79) ~[zip4j-2.2.4.jar:na]

**Comment by srikanth-lingala** [comment]
A sample zip file would help, but I am not sure where you can upload a file of 1.6GB. Maybe if you have a google account, you can upload it to google drive and share the link with me?

Were there multiple files in the zip file or was it a one file of such size?

**Comment by jrbocarsly** [comment]
A couple of additional notes:
1) You can download the sample zip (1.6 GB) from:
https://drive.google.com/file/d/1ITmZZ1TFFxg7DGPEghzrwDDOt-sHLarh/view?usp=sharing

2. There are 6 files in the zip: 0 B, 616 B, 191 KB 48.5 MB, 1.3 GB, 7.1 GB

3. On building the zip, no exceptions are thrown

4. We see exceptions only on trying to expand the file (see trace above)

**Comment by srikanth-lingala** [comment]
Do you see the same behaviour also when the zip file was created with the latest version of zip4j (v2.6.3)?

**Comment by jrbocarsly** [comment]
Yes, unfortunately, we do see the same behavior with 2.6.3 - we tried upgrading right away.

Regards
Jeff



Sent from my Verizon, Samsung Galaxy smartphone



-------- Original message --------

**Comment by srikanth-lingala** [comment]
The zip file is password protected. Would it be possible to share the password for me to test? If you can, but do not want to share here publicly, you can also send me an email to srikanth.mailbox@gmail.com

**Comment by jrbocarsly** [comment]
Apologies, I forgot about that. I’ll mail you the password – however, note that all the data is generated and has no significance outside this example.

**Comment by srikanth-lingala** [comment]
I see something strange with the sql files in your zip file. The files are marked as having compression (and not encryption) type AES (which is set in case of aes encryption), but the file encryption is set as zip_standard. 

<img width="615" alt="Bildschirmfoto 2020-10-15 um 19 27 17" src="https://user-images.githubusercontent.com/3450302/96165727-47eb4000-0f1d-11eb-9415-cfadc99fb38b.png">

How are you creating the zip file? Can you post your code please? Are you in any way modifying the zip file after it has been created?

**Comment by srikanth-lingala** [comment]
Another quesion: do you see this behaviour on every zip file generated from your code? Or is it only with some specific files?

**Comment by jrbocarsly** [comment]
Srikanth

1)      Our process creates the zip, streams files to it, the goes through another few cycles of adding files to it, creating a new stream each time (perhaps that is the issue?). All streams are created with the same params (see below).

2)      The issue occurs only with some zip files. It appears to be related to the size of the zip, since we mainly see it in larger zips, but it isn’t clear to us how that works, i.e. is it the size of one file in the zip? Is it the total size? Etc.

Regards
Jeff

ZipParameters zipParameters = new ZipParameters()
zipParameters.setCompressionMethod(CompressionMethod.DEFLATE);
zipParameters.setEncryptionMethod(EncryptionMethod.AES);
zipParameters.setAesKeyStrength(AesKeyStrength.KEY_STRENGTH_256);

We add files with plain vanilla code:
              if (pwd != null) {
                     zipParameters.setEncryptFiles(true);
                     zipFile = new ZipFile(outputZipTargetFile, pwd.toCharArray());
              } else {
                     zipParameters.setEncryptFiles(false);
                     zipFile = new ZipFile(outputZipTargetFile);
              }

              for (File fileToAdd : filesToAdd) {
                     try (InputStream inputStream = new FileInputStream(fileToAdd)) {
                           zipParameters.setFileNameInZip(fileToAdd.getName());
                           zipFile.addStream(inputStream, zipParameters);
                     }
           }

Pl

**Comment by srikanth-lingala** [comment]
Unfortunately all my attempts to reproduce this issue with a code similar to yours has failed. I tried to generate a zip file approximately the same size as yours, and in fact, I tried generating even bigger zip file, and I had a valid zip file every time. I do not have the same files that are part of your zip file as I am not able to extract the zip file. I generated random files of a few GB and tried adding them to the zip file via streams. Here is the code I tried:

```
@Test
public void testIssue248() throws IOException {
  ZipFile zipFile = new ZipFile(new File(BASE_DIR + "test_me.zip"), "password".toCharArray());
  List<File> filesToAdd = Arrays.asList(
      new File(BASE_DIR + "extract/qsdb_00000000000000000000000000000001_d963cca6a41846c7a2d1ff678a79d4d1.sql"),
      new File(BASE_DIR + "extract/QUERYSURGE_PROJECT_EXPORT_META_d963cca6a41846c7a2d1ff678a79d4d1.MD"),
      new File(BASE_DIR + "extract/20201008_092535_WOUTPWD")
  );
  ZipParameters zipParameters = new ZipParameters();
  zipParameters.setEncryptFiles(true);
  zipParameters.setEncryptionMethod(EncryptionMethod.AES);
  
  zipFile.addFiles(filesToAdd);

  List<File> filesToAddLater = Arrays.asList(
      new File(BASE_DIR + "output/FILE_2"), // about 2 GB
      new File(BASE_DIR + "output/FILE_3") // about 1 GB
  );

  for (File fileToAdd : filesToAddLater) {
    zipParameters.setFileNameInZip(fileToAdd.getName());
    zipFile.addStream(new FileInputStream(fileToAdd), zipParameters);
  }
}
```

I really wonder where and why the encryption method changes from aes to zip standard in your case. 

A few more questions:

1.  What happens if you try to add the file directly instead of adding a stream. In your code snippet, instead of:

```
for (File fileToAdd : filesToAdd) {
         try (InputStream inputStream = new FileInputStream(fileToAdd)) {
               zipParameters.setFileNameInZip(fileToAdd.getName());
               zipFile.addStream(inputStream, zipParameters);
         }
}
 ```

directly adding the list of files to zip:

```
zipFile.addFiles(filesToAdd, zipParameters);
```

2. With the set of files in the zip file you uploaded, are you able to reproduce the issue each time you create a zip file?

3. When you said earlier that you tried with the latest version of zip4j, did you also create the zip file with the latest version or did you just extract the zip file with the latest version?

I suspect there is something else happening in your code that is corrupting the files in the zip. This is definitely a bug in zip4j, and I am trying hard to figure out the circumstances leading to zip4j generating a corrupted zip file. All I want is to reproduce this issue once, and I can fix it. I tried several times with several different file sizes, and each time I end up with a valid zip file.

**Comment by jrbocarsly** [comment]
Per your numbering above:

1) We'll give this approach a try and get back to you

2) Our impression is that the zip files we're creating in this size range (and presumably, above this range) all have the problem, but what the size cutoff is or why this happens at a given size is unknown to us. We've tried smaller sizes (~0.25 GB) and these appear to work fine (and nothing in between the two sizes unfortunately).

3) I think we only tried to open the existing zip (created in 2.2.4) with 2.6.3; we can try the whole process with 2.6.3

Additionally, we'll see if we can get a set of our "pre-zipped" files for you to work with. This will take a bit of time on our side to set up (hopefully not more than a couple days). 

I'm wondering whether the issue has something to do with the fact that we're adding the files in steps, with a fresh ZipFile ref each time? Again, it works with significantly smaller sizes.

Thanks
Jeff

**Comment by srikanth-lingala** [comment]
I tried with fresh ref to zip file each time as well, but again could not reproduce the issue. It would be really awesome if you could provide me with a reproducible package (code + files part of the zip file). I will give it a few more tries and will let you know if I manage to reproduce it. Unless I reproduce this issue, it will be pretty hard to figure out what/where is going wrong. Also, I would be eagerly waiting for the results of point 1 (adding files with addFiles()) and 3 (whole process with 2.6.3) from your post above.

**Comment by jrbocarsly** [comment]
Well...upgrading to v.2.6.3 is something we should have tried first...we built a test rig that reproduces the issue using v.2.2.4. When we switched over to v.2.6.3, and the zip files were intact. We've verified this out in our product - we get a valid zip. 

We were mislead by 2.6.3 not being able to open what we now know was a "broken" zip created in 2.2.4. So it appears that, as regards this issue, we're in good shape.

Many thanks for your quick response and the helpful discussion on this - much appreciated.

Regards
Jeff
